### PR TITLE
Geometry template and barrel extension

### PR DESCRIPTION
**Pull request by ebrondol**
This PR aims to address the following open issues:
- Fix https://github.com/key4hep/k4Clue/issues/25 
- Fix https://github.com/key4hep/k4Clue/issues/24

BEGINRELEASENOTES
- Template CLUEAlgo and CLUEAlgoGPU on TILE_CONST
- Correct treatment of different part of the ECAL detector: EB, EE+ and EE-
- Phi search extension in CLUEAlgo
ENDRELEASENOTES

Still to do to finalize this PR:
- [x] More clear readme.md
- [x] Recipe on how to add a new detector
- [x] Clean up code in several places
- [x] Template on the `TILE` and insert `constants_type_t` to access to the variables in the `TILE_CONST`
~~Macro folder also to be cleaned up and make accessible to other developers~~ -> keep it for a later PR

Developments on GPU:
- Possible to test after https://github.com/key4hep/k4Clue/issues/34 was closed
- Correctly passing the compilation and the test
- The output file do not contains clusters, but only outliers, this will be fixed in a subsequent PR (Open issue https://github.com/key4hep/k4Clue/issues/37)

@vvolkl

### CODE REVIEW

**Comment by vvolkl** [comment]
With the added suffix, `LayerTilesGPUT` becomes a bit hard to read, maybe better `LayerTilesGPU_T`?

Otherwise LGTM.

**Comment by ebrondol** [comment]
From my side this PR is ready. 

The tests (and fixes, if needed) on the GPU part can also be done in another PR when [the temporary fix on the initializer list for nvcc](https://github.com/AIDASoft/podio/pull/333) is merged.

**Comment by ebrondol** [comment]
@rovere might also want to send a review of this PR before merging

**Comment by ebrondol** [comment]
Hi @vvolkl @tmadlener ,
from my side this PR is ready to go in.
I have reported the situation on the GPU side in the description of the PR but given that the tests are running fine and the compilation as well, I would postpone this to a future PR.

**Comment by ebrondol** [comment]
Hi @vvolkl @tmadlener 
Any news on the merging of this PR ?
Happy 2023!

File: include/LayerTiles.h [hunk 1]
```diff
@@ -65,6 +87,23 @@ class LayerTiles {
       return std::array<int, 4>({{ xBinMin,xBinMax,yBinMin,yBinMax }});
     }
 
+    std::array<int, 4> searchBoxPhiZ(float phiMin, float phiMax, float zMin, float zMax) const {
+      int phiBinMin = getPhiBin(phiMin);
+      int phiBinMax = getPhiBin(phiMax);
+      // If the search window cross the phi-bin boundary, add T::nPhiBins to the
+      // MAx value. This guarantees that the caller can perform a valid doule
```
Comment (andresailer) [inline include/LayerTiles.h:94]:
```suggestion
      // Max value. This guarantees that the caller can perform a valid double
```

File: include/LayerTiles.h [hunk 2]
```diff
@@ -65,6 +87,23 @@ class LayerTiles {
...
+      // loop on eta and phi. It is the caller responsibility to perform a module
```
Comment (andresailer) [inline include/LayerTiles.h:95]:
```suggestion
      // loop on eta and phi. It is the caller responsibility to perform a modulo
```

File: include/LayerTiles.h [hunk 3]
```diff
@@ -65,6 +87,23 @@ class LayerTiles {
...
+      // operation on the phiBin values returned by this function, to explore the
+      // correct bins.
```
Comment (andresailer) [inline include/LayerTiles.h:97]: This should be a doxygen style comment that would eventually be visible in the documentation of the source code

File: src/CLUEAlgo.cc [hunk 4]
```diff
@@ -86,18 +112,27 @@ void CLUEAlgo::calculateDistanceToHigher( std::array<LayerTiles, NLAYERS> & allL
     int nearestHigher_i = -1;
     float xi = points_.x[i];
     float yi = points_.y[i];
+    float ri = points_.x[i]/points_.phi[i];
```
Comment (andresailer) [inline src/CLUEAlgo.cc:115]: `phi[i]` could be zero? What value does this calculate? (though later you divide by `ri` so you could calculate the inverse here and then later multiply, although then `x[i]` could be zero?)

File: src/CLUENtuplizer.cpp [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+  }
+
+  if (service("THistSvc", m_ths).isFailure()) {
+    error() << "Couldn't get THistSvc" << endmsg;
+    return StatusCode::FAILURE;
+  }
+
+  t_hits = new TTree ("CLUEHits", "CLUE calo hits ntuple");
+  if (m_ths->regTree("/rec/NtuplesHits", t_hits).isFailure()) {
+    error() << "Couldn't register hits tree" << endmsg;
```
Comment (andresailer) [inline src/CLUENtuplizer.cpp:33]: Do the errors() mean one should `return StatusCode::FAILURE` ?

File: src/CLUENtuplizer.cpp [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+    m_clusters_z->push_back (cl.getPosition().z);
+
+    // Sum up energy of cluster hits and save info
+    // Printout the hits that are in Ecal but not included in the clusters
+    int maxLayer = 0;
+    for (const auto& hit : cl.getHits()) {
+      foundInECAL = false;
+      for (const auto& clEB : *EB_calo_coll) {
+          if( clEB.getCellID() == hit.getCellID()){
+            foundInECAL = true;
```
Comment (andresailer) [inline src/CLUENtuplizer.cpp:133]:
```suggestion
            foundInECAL = true;
            break;  // Found in EB, break the loop
```
No need for the next if

File: src/ClueGaudiAlgorithmWrapper.cpp [hunk 7]
```diff
@@ -106,25 +145,21 @@ void ClueGaudiAlgorithmWrapper::fillFinalClusters(const std::map<int, std::vecto
...
-          if( index < EB_calo_coll->size() ) {
-            cluster.addToHits(EB_calo_coll->at(index));
-          } else {
-            cluster.addToHits(EE_calo_coll->at(index - EB_calo_coll->size()));
-          }
-        } else {
+
+        position.x += clue_hits[index].getPosition().x;
+        position.y += clue_hits[index].getPosition().y;
+        position.z += clue_hits[index].getPosition().z;
```
Comment (andresailer) [inline src/ClueGaudiAlgorithmWrapper.cpp:153]: is position used somewhere?

**Review by andresailer — COMMENTED** [review]
Generic Comment
* move the header files to `include/clue` and #include "clue/...", to separate things a little bit more in common insallations or views
* Otherwise also mostly minor comments, nothing blocking, I think.


Thanks!

File: src/ClueGaudiAlgorithmWrapper.cpp (hunk 7 above)
Comment (ebrondol) [reply src/ClueGaudiAlgorithmWrapper.cpp:153]: after we implemented the position re-weighted in log, no!
Thanks for the catch!

File: src/CLUEAlgoGPU.cu [hunk 8]
```diff
@@ -91,19 +92,20 @@ __global__ void kernel_calculate_distanceToHigher(LayerTilesGPU* d_hist,
     float rhoi = d_points.rho[i];
 
     // get search box 
-    int4 search_box = d_hist[layeri].searchBox(xi-dm, xi+dm, yi-dm, yi+dm);
+    auto lt = d_hist[layeri];
```
[the remark this answers is not in the document]
Comment (ebrondol) [reply src/CLUEAlgoGPU.cu:95]: No, this change gives the following error:
```
/afs/cern.ch/user/e/ebrondol/work/clue/k4Clue/src/CLUEAlgoGPU.cu(96): error: the object has type qualifiers that are not compatible with the member function "LayerTilesGPU_T<T>::searchBox [with T=LayerTilesConstants]"
            object type is: const clue::LayerTileGPU
          detected during:
            instantiation of "void kernel_calculate_distanceToHigher(TILES &, PointsPtr, float, float, int) [with TILES=LayerTilesGPU]" 
(222): here
            instantiation of "void CLUEAlgoGPU_T<TILES>::makeClusters() [with TILES=LayerTilesGPU]" 
(238): here
```

File: include/LayerTilesConstants.h [hunk 9]
```diff
@@ -1,30 +1,29 @@
...
 #define LayerTilesConstants_h
 
-#include <cstdint>
+#include "constexpr_cmath.h"
 #include <array>
 
+struct LayerTilesConstants {
 
-#define NLAYERS 1000
+  static constexpr float minX =  -250.f;
```
[the remark this answers is not in the document]
Comment (ebrondol) [reply include/LayerTilesConstants.h:12]: I have included 
```
// Variables meaning is specified in
// include/readme.md
```
in the header. I hope this helps people to find the information they need.

**Comment by ebrondol** [comment]
Hi @andresailer, with regards to the suggestion here below (and other similar ones):

> Generic Comment
> 
> * move the header files to `include/clue` and #include "clue/...", to separate things a little bit more in common insallations or views

We were in an ongoing discussion with @vvolkl about introducing [kalos/clue](https://gitlab.cern.ch/kalos/clue) in the key4hep stack so that we could have taken the entire library instead of copy-paste. I would keep this library independent from DD4hep geometry, for example. This is still under discussion and I hope it will be implemented soon.

**Comment by ebrondol** [comment]
Hi @andresailer , if I am not mistaken, I have resolved all comments.
The one related to the possibility of `phi` being zero needed some changes in the code. It was also resolved today and tested. Please have a look if I missed anything - thank you.

**Review by andresailer — APPROVED** [review]

File: docs/clic-recipe.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+To generate particle from the suface of the CLICdet ECAL, substitute the generation command with the following
+```bash
+ddsim --steeringFile clic_steer.py --compactFile $LCGEO/CLIC/compact/CLIC_o3_v14/CLIC_o3_v14.xml --enableG4GPS --runType "run" --macroFile  myGPS.mac --outputFile gps_gamma_10GeV_edm4hep.root
+```
+where `myGPS.mac` can be found in this folder.
+
+
+## Visualization
+
+If you want to visualise the output as event display using CED:
```
Comment (andresailer) [inline docs/clic-recipe.md:44]:
```suggestion
If you want to visualise the output with the CED event display:
```
(Just some change to hopefully trigger the CI build test...)

**Pull request merged**